test/cadet_web/controllers/assessments_controller_test.exs

```diff
@@ -1,11 +1,378 @@
+
+        # Seeds set solution as 0
+        expected_mcq_solutions = Enum.map(mcq_questions, fn _ -> %{"solution" => 0} end)
+
+        expected_programming_solutions =
+          Enum.map(programming_questions, &%{"solution" => &1.question.solution})
+
+        expected_solutions = Enum.sort(expected_mcq_solutions ++ expected_programming_solutions)
+
+        resp_solutions =
+          conn
+          |> sign_in(user)
+          |> get(build_url(assessment.id))
+          |> json_response(200)
+          |> Map.get("questions")
+          |> Enum.map(&Map.take(&1, ["solution"]))
+          |> Enum.sort()
+
+        assert expected_solutions == resp_solutions
+      end
+    end
+
+    test "it does not render solutions for ungraded assessments (path)", %{
+      conn: conn,
+      users: users,
+      assessments: assessments
+    } do
+      for role <- Role.__enum_map__() do
+        user = Map.get(users, role)
+
+        for {_type,
+             %{
+               assessment: assessment
+             }} <- Map.delete(assessments, :path) do
+          resp_solutions =
+            conn
+            |> sign_in(user)
+            |> get(build_url(assessment.id))
+            |> json_response(200)
+            |> Map.get("questions")
+            |> Enum.map(&Map.get(&1, ["solution"]))
+
+          assert Enum.uniq(resp_solutions) == [nil]
+        end
+      end
+    end
+
+    test "it does not permit access to not yet open assessments", %{
+      conn: conn,
+      users: users,
+      assessments: %{mission: mission}
+    } do
+      for role <- Role.__enum_map__() do
+        user = Map.get(users, role)
+
+        {:ok, _} =
+          mission.assessment
+          |> Assessment.changeset(%{
+            open_at: Timex.shift(Timex.now(), days: 5),
+            close_at: Timex.shift(Timex.now(), days: 10)
+          })
+          |> Repo.update()
+
+        conn =
+          conn
+          |> sign_in(user)
+          |> get(build_url(mission.assessment.id))
+
+        assert response(conn, 401) == "Assessment not open"
+      end
+    end
+
+    test "it does not permit access to unpublished assessments", %{
+      conn: conn,
+      users: users,
+      assessments: %{mission: mission}
+    } do
+      for role <- Role.__enum_map__() do
+        user = Map.get(users, role)
+
+        {:ok, _} =
+          mission.assessment
+          |> Assessment.changeset(%{is_published: false})
+          |> Repo.update()
+
+        conn =
+          conn
+          |> sign_in(user)
+          |> get(build_url(mission.assessment.id))
+
+        assert response(conn, 400) == "Assessment not found"
+      end
+    end
+  end
+
+  describe "GET /assessment_id, student" do
+    test "it renders previously submitted answers", %{
+      conn: conn,
+      users: %{student: student},
+      assessments: assessments
+    } do
+      for {_type,
+           %{
+             assessment: assessment,
+             mcq_answers: [mcq_answers | _mcq_answers],
```

Comment:
It's ok to write this as `[mcq_answers | _]` tbh cos I know you're discarding the tail haha

Reply:
Changed